crates/iota-names/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+pub mod config;
+pub mod domain;
+pub mod error;
+pub mod registry;
+
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use iota_types::base_types::ObjectID;
+use move_core_types::{
+    account_address::AccountAddress, ident_str, identifier::IdentStr, language_storage::StructTag,
+};
+use serde::{Deserialize, Serialize};
+
+use self::domain::Domain;
+
+pub const MIN_LABEL_LEN: usize = 3;
+pub const MAX_LABEL_LEN: usize = 63;
+
+/// An object to manage a second-level domain (SLD)
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct IotaNamesRegistration {
+    id: ObjectID,
+    domain: Domain,
+    domain_name: String,
+    expiration_timestamp_ms: u64,
+    image_url: String,
+}
+
+/// An object to manage a subdomain.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct SubdomainRegistration {
+    id: ObjectID,
+    nft: IotaNamesRegistration,
+}
+
+impl SubdomainRegistration {
+    pub fn into_inner(self) -> IotaNamesRegistration {
+        self.nft
+    }
+}
+
+/// Unifying trait for [`IotaNamesRegistration`] and [`SubdomainRegistration`]
+pub trait IotaNamesNft {
+    const MODULE: &IdentStr;
+    const TYPE_NAME: &IdentStr;
+
+    fn type_(package_id: AccountAddress) -> StructTag {
+        StructTag {
+            address: package_id,
+            module: Self::MODULE.into(),
+            name: Self::TYPE_NAME.into(),
+            type_params: Vec::new(),
+        }
+    }
+
+    fn domain(&self) -> &Domain;
+
+    fn domain_name(&self) -> &str;
+
+    fn expiration_timestamp_ms(&self) -> u64;
+
+    fn expiration_time(&self) -> SystemTime;
+
+    fn has_expired(&self) -> bool {
+        self.expiration_time() <= SystemTime::now()
+    }
+
+    fn image_url(&self) -> &str;
+
+    fn id(&self) -> ObjectID;
+}
+
+impl IotaNamesNft for IotaNamesRegistration {
+    const MODULE: &IdentStr = ident_str!("iota_names_registration");
+    const TYPE_NAME: &IdentStr = ident_str!("IotaNamesRegistration");
+
+    fn domain(&self) -> &Domain {
+        &self.domain
+    }
+
+    fn domain_name(&self) -> &str {
+        &self.domain_name
+    }
+
+    fn expiration_timestamp_ms(&self) -> u64 {
+        self.expiration_timestamp_ms
+    }
+
+    fn expiration_time(&self) -> SystemTime {
+        UNIX_EPOCH + Duration::from_millis(self.expiration_timestamp_ms())
```

Comment:
Oh I thought the expiration_timestamp_ms is already in unix time? The UNIX_EPOCH here confuses me a bit. So it's not? Or is that only needed to make a SystemTime object out of it?

Reply:
Not sure to follow, it is a unix timestamp, which is the number os s/ms since the unix epoch, so if you want the actual date you need to add both

Reply:
yeah never mind.
the `expiration_timestamp_ms()` are the milliseconds since UNIX_EPOCH.
for creating a valid `SystemTime` you need to parse the duration from those millis and add the `UNIX_EPOCH`.
I was just a bit confused while reading it, because I never handled unix timestamps in rust yet.